### PR TITLE
cudnn based network eval for lc0

### PR DESCRIPTION
**Pull request by ankan-ban**
- About 50% faster than TF implementation.
- Much easier to build for windows (just install latest cuda sdk, and cudnn library, build using Visual studio project)
 - Compile time switch can be used to switch back to TF version.

On GTX 970 Benchmarked using:
setoption name Minibatch size for NN inference value 256
ucinewgame
isready
go nodes 100000

cudnn:  nps 10606
TF: nps 6819

### CODE REVIEW

File: lc0/src/neural/network_cudnn.cu [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+public:
+    SoftMaxLayer(BaseLayer *ip);
+    void eval(void *output, void *input, void *input2, void *scratch) override;
+};
+
+class BNLayer : public BaseLayer
+{
+private:
+    void *means;
+    void *variences;
```
Comment (dubslow) [inline lc0/src/neural/network_cudnn.cu]: spelling, this should be "variances"

File: lc0/src/neural/network_cudnn.cu [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+        fixBiases(weights.policy);
+        fixBiases(weights.value);
+        fixStdDevs(weights.policy.bn_stddivs);
+        fixStdDevs(weights.value.bn_stddivs);
+
+
+
+
+        // 1. build the network, and copy the weights to GPU memory
+        int weightIndex = 0;
```
[the remark this answers is not in the document]
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: fixed

File: lc0/src/neural/network_cudnn.cu (hunk 1 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: fixed.

File: lc0/LC0VSProj/LC0VSProj.vcxproj [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
+    <LinkIncremental>true</LinkIncremental>
+    <IncludePath>$(ProjectDir)\..\src;$(VC_IncludePath);$(WindowsSDK_IncludePath);E:\Ankan\MachineLearning\CUDA_NN\7.1</IncludePath>
```
Comment (Akababa) [inline lc0/LC0VSProj/LC0VSProj.vcxproj]: remove hardcoding?

**Comment by Ttl** [comment]
You can multiple convolution weights by bn stddivs and move bn means to convolutions biases to get rid of the batch norm layer. It should be little bit faster especially with small batch sizes. (Implementation here: https://github.com/Ttl/leela-zero/commit/842d0cb61fedc47fa9012fbec00608fbb0e7cf67)

Also here's my cudnn branch for LZGo if you are interested: https://github.com/Ttl/leela-zero/commit/11a254cf06a5577abcf6de97fdc2d72de90d0d62.

**Comment by mooskagh** [comment]
Could you also enable it for Linux?
For that, just replace meson.build with https://gist.github.com/mooskagh/85d9ccc0c509ab239879d63b672e9a13
and it works.

And then it's fine to remove all #ifdefs for now, I'm implementing proper multi-backend support anyway.

**Comment by ankan-ban** [comment]
updated meson build from mooskagh and made cudnn path default on all platforms. 
I still have #ifdef at some places to make it easier for people to build just the cudnn version (without bothering about tensorflow dependencies).

**Comment by ankan-ban** [comment]
also implemented suggestion by @Ttl of getting rid of batch-norm layer by adjusting weights/biases of convolutions.
 - nps 11454 (~7% better than before)

**Comment by Error323** [comment]
Awesome! Have the network results been verified against the tensorflow version? Perhaps we should setup a unittest for this to sanity check the results of a simple random network against various backends.

**Comment by ankan-ban** [comment]
As I didn't attempt to build the tensorflow version, I only verified it by comparing the PV outputs. 

Initially for debugging, I had manually tested outputs of a few board evals comparing it with the output of main branch (openCL/cpu version).

File: lc0/src/neural/network_cudnn.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "network.h"
+
+namespace lczero {
+
+// Creates a network which just returns random values (for plain nps testing).
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.h]: It doesn't just return random values. :)

File: lc0/src/neural/network_cudnn.cu [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+}
+
+#define reportCUDNNErrors(status)  cudnnError(status, __FILE__, __LINE__)
+
+
+// 256 MB fixed scratch memory size (hardcoded for now)
+#define CUDA_SCRATCH_SIZE (256*1024*1024)
+
+// hard-coded for now, no point in going above this anyway (can possibly save memory by reducing this)
+static constexpr int MAX_BATCH_SIZE = 1024;
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: I used google style guide for other C++ files. https://google.github.io/styleguide/cppguide.html
Would it be possible to change this file to use it too, so that file format is consistent.

Particularly:
Names of global constants are in format kMaxBatchSize.

Multi-word variable name are in this format: multi_word_name
Class data members have trailing underscore:  in_tensor_desc_

Most of other requirements can be fixed automatically by clang-format (I can run it later on checked-in code)

File: lc0/src/neural/network_cudnn.cu [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+        getchar();
+        exit(1);
+    }
+}
+
+#define reportCUDNNErrors(status)  cudnnError(status, __FILE__, __LINE__)
+
+
+// 256 MB fixed scratch memory size (hardcoded for now)
+#define CUDA_SCRATCH_SIZE (256*1024*1024)
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: It's better to have C++ constants rather than macros for macros. (unless CUDA requires it so)
Also it's encouraged to put them into anonymous namespace rather than make static.

File: lc0/src/neural/network_cudnn.cu [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+
+#include <cublas_v2.h>
+#include <cudnn.h>
+
+
+void cudnnError(cudnnStatus_t status, const char* file, const int& line)
+{
+    if (status != CUDNN_STATUS_SUCCESS)
+    {
+        printf("CUDNN error: %s (%s:%d)\n", cudnnGetErrorString(status), file, line);
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Consider throwing an exception (from utils/exception.h) instead, that way UCI loop has a chance to recover.

File: lc0/src/neural/network_cudnn.cu [hunk 8]
```diff
@@ -0,0 +1,836 @@
...
+//  1. input for the layer
+//  2. output of the layer
+//  3. data from old layer for skip connection
+
+
+/////////////////////////////////////////////////////////////////////////////
+//                      Global/Static variables                            //
+/////////////////////////////////////////////////////////////////////////////
+
+int BaseLayer::N = 1;
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: I believe that if you make them constexpr, you can initialize them right in a class definition.

File: lc0/src/neural/network_cudnn.cu [hunk 9]
```diff
@@ -0,0 +1,836 @@
...
+bool BaseLayer::fp16 = false;
+size_t BaseLayer::bpe = sizeof(float);
+
+cudnnHandle_t g_cudnn;
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Is there a reason those are global and not CudnnNetwork members?
When running code on two GPUs, we'll have to have two separate copies of them, right?

File: lc0/src/neural/network_cudnn.cu [hunk 10]
```diff
@@ -0,0 +1,836 @@
...
+
+    cudnnSetConvolution2dDescriptor(convDesc,
+        padding, padding,
+        1, 1,
+        1, 1,
+        crossCorr ? CUDNN_CROSS_CORRELATION : CUDNN_CONVOLUTION,
+        fp16 ? CUDNN_DATA_HALF : CUDNN_DATA_FLOAT);
+
+    /*
+    cudnnGetConvolutionForwardAlgorithm(g_cudnn,
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Probably not worth keeping it here.

File: lc0/src/neural/network_cudnn.cu [hunk 11]
```diff
@@ -0,0 +1,836 @@
...
+}
+
+FCLayer::~FCLayer()
+{
+    cudaFree(weights);
+    cudaFree(biases);
+}
+
+
+namespace lczero {
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Would it be possible to put all the locally defined classes (above) into the anonymous namespace?
I'm not sure if that's compatible with CUDA compiler, but if it is, that's surely something to do, in order not to clutter global namespace.

Also if that's possible, consider moving this line to the very top of the file.

File: lc0/src/neural/network_cudnn.cu [hunk 12]
```diff
@@ -0,0 +1,836 @@
...
+     // output (TODO: try using cudaHostAlloc to avoid the copy?)
+     float out_pol[MAX_BATCH_SIZE][V2_NUM_OUTPUT_POLICY];
+     float out_val[MAX_BATCH_SIZE];
+     float input_planes[MAX_BATCH_SIZE][kInputPlanes*8*8];
+
+     const CudnnNetwork* m_network;
+};
+
+// currently only one NN eval can happen a time (we can fix this if needed by allocating more memory)
+static std::mutex g_lock;
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: That is also something that seems, that it should be inside CudnnNetwork class.

File: lc0/src/neural/network_cudnn.cu [hunk 13]
```diff
@@ -0,0 +1,836 @@
...
+        // 3. allocate scratch space (used internally by cudnn to run convolutions)
+        cudaMalloc(&scratchMem, CUDA_SCRATCH_SIZE);
+
+    }
+
+    void forwardEval(float *input, float *op_pol, float *op_val, int batchSize) const
+    {
+        // printf(" ..%d.. ", batchSize);
+
+        g_lock.lock();
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Consider using std::lock_guard instead of locking and unlocking manually.

File: lc0/src/neural/network_cudnn.cu [hunk 14]
```diff
@@ -0,0 +1,836 @@
...
+        network[l++]->eval(tensorMem[0], tensorMem[2], NULL, scratchMem);    // value FC1
+        network[l++]->eval(tensorMem[2], tensorMem[0], NULL, scratchMem);    // value FC2    // VALUE
+
+        // copy results back to CPU memory
+        cudaMemcpyAsync(&op_pol[0], tensorMem[1], policyOut->getOutputSize(), cudaMemcpyDeviceToHost);
+        cudaError_t status = cudaMemcpy(&op_val[0], tensorMem[2], valueOut->getOutputSize(), cudaMemcpyDeviceToHost);
+
+        if (status != cudaSuccess)
+        {
+            printf("\nSome error running cuda based eval!\n");
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: Consider changing to either assert of Exception.

File: lc0/src/neural/network_cudnn.cu (hunk 6 above)
Comment (Ttl) [reply lc0/src/neural/network_cudnn.cu]: This should probably be calculated to avoid wasting GPU memory. See cudnnGetConvolutionForwardWorkspaceSize function.

File: lc0/src/neural/network_cudnn.cu (hunk 5 above)
Comment (Ttl) [reply lc0/src/neural/network_cudnn.cu]: 1024 seems very big for maximum batch size. I don't think low end cards even have enough memory to fit this many batches. This too should be probably allocated dynamically to avoid wasting huge amounts of GPU memory.

File: lc0/src/neural/network_cudnn.cu (hunk 9 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: correct, fixed.

File: lc0/src/neural/network_cudnn.cu (hunk 11 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: moved everything inside lczero namespace

File: lc0/src/neural/network_cudnn.cu (hunk 5 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: updated most names, except for ones in small util classes (layer implementations).

File: lc0/src/neural/network_cudnn.cu (hunk 8 above)
Comment (mooskagh) [reply lc0/src/neural/network_cudnn.cu]: Then it should not be static. We'll need to support running it on multiple GPUs soon, and having it static will make it harder.

File: lc0/src/neural/network_cudnn.cu (hunk 14 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: done.

File: lc0/src/neural/network_cudnn.cu (hunk 13 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: done

File: lc0/src/neural/network_cudnn.cu (hunk 12 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: right. done.

File: lc0/src/neural/network_cudnn.cu (hunk 8 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: fixed. Made it a parameter to eval.

File: lc0/src/neural/network_cudnn.cu (hunk 7 above)
Comment (ankan-ban) [reply lc0/src/neural/network_cudnn.cu]: done.

File: lc0/LC0VSProj/LC0VSProj.vcxproj (hunk 3 above)
Comment (ankan-ban) [reply lc0/LC0VSProj/LC0VSProj.vcxproj]: done, but the developer will still need to make it point to location where cudnn library is present (no standard define for cudnn path :-/).

File: .gitignore [hunk 15]
```diff
@@ -32,3 +32,4 @@ training/tf/models/
 /go/src/client/settings.json
 *.vspx
 *.psess
+/lc0/LC0VSProj/*
```
Comment (mooskagh) [inline .gitignore]: There is a local .gitignore at ./lc0, move it there.

File: lc0/src/neural/network_cudnn.cu [hunk 16]
```diff
@@ -0,0 +1,822 @@
...
+        network_[l++]->eval(batchSize, tensor_mem_[0], tensor_mem_[2], NULL, scratch_mem_, cudnn_, cublas_);    // value FC1
+        network_[l++]->eval(batchSize, tensor_mem_[2], tensor_mem_[0], NULL, scratch_mem_, cudnn_, cublas_);    // value FC2    // VALUE
+
+        // copy results back to CPU memory
+        cudaMemcpyAsync(&op_pol[0], tensor_mem_[1], policy_out_->getOutputSize(batchSize), cudaMemcpyDeviceToHost);
+        cudaError_t status = cudaMemcpy(&op_val[0], tensor_mem_[2], value_out_->getOutputSize(batchSize), cudaMemcpyDeviceToHost);
+
+        if (status != cudaSuccess)
+        {
+            throw(new Exception("Some error running cuda based eval!"));
```
Comment (mooskagh) [inline lc0/src/neural/network_cudnn.cu]: No need for `new`, just `throw Exception("blah")`

**Comment by mooskagh** [comment]
Thanks!
